lib/kafka_producer.js

```diff
@@ -157,10 +244,54 @@ KafkaProducer.prototype.getWholeMsg = function getWholeMsg(topic, message, timeS
 
 KafkaProducer.prototype.close = function close(callback) {
     var self = this;
+    if (self.batching) {
+        self.flushEntireCache(callback);
+    }
     self.enable = false;
     if (self.restClient) {
         self.restClient.close();
     }
 };
 
+// The batch message payload should follow:
+// 4 BE bytes number of messages + 4 BE bytes size of message + actual message
+function MessageBatch(topic) {
+    var self = this;
+
+    self.payload = null;
+    self.timestamp = new Date().getTime();
+    self.numMessages = 0;
+    self.sizeBytes = 0;
+    self.topic = topic;
+}
+
+MessageBatch.prototype.addMessage = function addMessage(message) {
+    var self = this;
+
+    var messageBinary = new Buffer(message);
+    var sizeBuf = new Buffer(4);
+    sizeBuf.writeInt32BE(messageBinary.length, 0);
+
+    if (self.payload) {
+        self.payload = Buffer.concat([self.payload, sizeBuf, messageBinary]);
+    } else {
+        self.payload = Buffer.concat([sizeBuf, messageBinary]);
+    }
+
+    self.numMessages += 1;
+    self.sizeBytes += messageBinary.length;
+};
+
+MessageBatch.prototype.getBatchedMessage = function getBatchedMessage() {
+    var self = this;
+
+    var sizeBuf = new Buffer(4);
```

Comment:
To avoid unnecessary buffer allocations or copies we are going to backfill the length here.

```
self.cacheBuf.writeUInt32BE(self.numMessages, 0)
```

This is safe to do because we started the offset at `4` in the constructor and thus left space to write this at the start.